mlir-sycl/lib/Dialect/IR/SYCLOps.cpp

```diff
@@ -98,19 +98,31 @@ mlir::LogicalResult mlir::sycl::SYCLAccessorSubscriptOp::verify() {
     return emitOpError("Dimensions cannot be zero");
 
   const auto verifyResultType = [&]() -> mlir::LogicalResult {
-    const auto resultType = getResult().getType().dyn_cast<mlir::MemRefType>();
+    const auto ResultType = getResult().getType();
+
+    return TypeSwitch<mlir::Type, mlir::LogicalResult>(ResultType)
+        .Case<mlir::MemRefType,
+              LLVM::LLVMPointerType>([&](auto Ty) -> mlir::LogicalResult {
+          const mlir::Type ElemType = Ty.getElementType();
+          if (ElemType != AccessorTy.getType())
+            return emitOpError(
+                       "Expecting a reference to this accessor's value type (")
+                   << AccessorTy.getType() << "). Got " << Ty;
+
+          if constexpr (std::is_same<decltype(Ty),
+                                     LLVM::LLVMPointerType>::value) {
+            if (!ElemType.isa<LLVM::LLVMStructType>())
+              return emitOpError(
+                         "Expecting pointer to struct return type. Got ")
+                     << ResultType;
```

Comment:
I would separate the Case into one for memref and one for LLVM pointer (given this code is only for LLVM ptr). 
Code at lines 106-110 would be duplicated, but you can make that into a lambda and call the lambda function in 2 places. Suggesting `CheckElemType` for the lambda name.